python/shotgun_data/background_task_manager.py

```diff
@@ -0,0 +1,832 @@
+                task_to_process = task
+                task_index = ti
+                break
+
+            if task_to_process:
+                # no need to look any further!
+                break
+
+        if not task_to_process:
+            # nothing to do!
+            return False
+
+        # we need a thread to do the work with:
+        thread = self._get_worker_thread()
+        if not thread:
+            # looks like we can't do anything!
+            return False
+
+        self._log("Starting task %r" % task_to_process)
+
+        # ok, we have a thread so lets move the task from the priority queue to the running list:
+        self._pending_tasks_by_priority[priority] = (self._pending_tasks_by_priority[priority][:task_index] 
+                                                     + self._pending_tasks_by_priority[priority][task_index+1:])
+        if not self._pending_tasks_by_priority[priority]:
+            # no more tasks with this priority so also clean up the list
+            del self._pending_tasks_by_priority[priority]
+        self._running_tasks[task_to_process.uid] = (task_to_process, thread)
+
+        num_tasks_left = 0
+        for pending_tasks in self._pending_tasks_by_priority.values():
+            num_tasks_left += len(pending_tasks)
+        self._log(" > Currently running tasks: '%s' - %d left in queue" % (self._running_tasks.keys(), num_tasks_left))
+
+        # and run the task
+        thread.run_task(task_to_process)
+
+        return True
+
+    def _on_task_completed(self, task, result):
+        """
+        Slot triggered when a task is completed.  This processes the result and emits the 
+        task_completed signal if needed.
+
+        :param task:    The task that completed
+        :param result:  The task result
+        """
+        try:
+            # check that we should process this result:
+            if task.uid in self._running_tasks:
+                self._log("Task %r - completed" % (task))
+
+                # if we have dependent tasks then update them:
+                for ds_task_id in self._downstream_task_map.get(task.uid) or []:
+                    ds_task = self._tasks_by_id.get(ds_task_id)
+                    if not ds_task:
+                        continue
+
+                    # update downstream task with result
+                    ds_task.append_upstream_result(result)
+
+                # remove the task:
+                group_finished = self._remove_task(task)
+
+                # emit signal that this task is completed:
+                self.task_completed.emit(task.uid, task.group, result)
+
+                if group_finished:
+                    # also emit signal that the entire group is completed:
+                    self.task_group_finished.emit(task.group)
+        finally:
+            # move this task thread to the available threads list:
+            self._available_threads.append(self.sender())
```

Comment:
Kinda violating encapsulation of the signal's source here. I think it would have been preferable if the QThread was obtained some other way. This will break in the future if for some reason an object that is not the thread invoked the slot.